cfgmgr/buffermgrdyn.cpp

```diff
@@ -2227,7 +3135,115 @@ void BufferMgrDynamic::doTask(Consumer &consumer)
+ *
+ * Originally, all buffer profiles and objects were applied to APPL_DB as soon as they were learnt from CONFIG_DB.
+ * The buffer profiles, and objects will be available as soon as they are received from CONFIG_DB when the system is starting
+ * but it takes more seconds for the buffer pools to be applied to APPL_DB
+ * because the sizes need to be calculated first and can be calculated only after some other data available.
+ * This causes buffer orchagent receives buffer profiles and objects first and then buffer pools.
+ * The buffer orchagent has to keep retrying in such case, which wastes time and incurs error messages.
+ *
+ * Now, we will guarantee the buffer profiles and objects to be applied only after buffer pools are ready.
+ * This is achieved by the following steps:
+ *  - m_bufferPoolReady indicates whether the buffer pools are ready.
+ *  - When the buffer profiles and objects are received from CONFIG_DB,
+ *    they will be handled and stored in internal data struct
+ *    (but not applied to APPL_DB until the buffer pools are ready, AKA. m_bufferPoolReady == true)
+ *  - Once buffer pools are ready, all stored buffer profiles and objects will be applied to APPL_DB by this function.
+ *
+ * The progress of pushing buffer pools into APPL_DB has also been accelerated by the lua plugin for calculating buffer sizes.
+ * Originally, if the data are not available for calculating buffer sizes, it just return an empty vectors and buffer manager will retry
+ * Now, the lua plugin will return the sizes calculated by the following logic:
+ *  - If there are buffer sizes in APPL_DB, return the sizes in APPL_DB.
+ *  - Otherwise, return all 0 as sizes.
+ * This makes sure there will be available sizes of buffer pools as sonn as possible when the system is starting
+ * The sizes will be updated to non zero when the data for calculating it are ready.
+ * Even this is Mellanox specific, other vendors can also take advantage of this logic.
+ */
+void BufferMgrDynamic::handlePendingBufferObjects()
+{
+    if (m_bufferPoolReady)
+    {
+        if (!m_pendingApplyZeroProfilePorts.empty())
+        {
+            // No action as no zero profiles defined.
+            if (!m_zeroPoolAndProfileInfo.empty() && !m_zeroProfilesLoaded)
+            {
+                loadZeroPoolAndProfiles();
+            }
+        }
+
+        if (m_bufferObjectsPending)
+        {
+            // Apply all pending buffer profiles
+            for (auto &profile : m_bufferProfileLookup)
+            {
+                updateBufferProfileToDb(profile.first, profile.second);
+            }
+
+            for (auto &port : m_portInfoLookup)
+            {
+                if (port.second.state != PORT_ADMIN_DOWN)
+                {
+                    refreshPgsForPort(port.first, port.second.effective_speed, port.second.cable_length, port.second.mtu);
+                    for (auto &pg : m_portPgLookup[port.first])
```

Comment:
looks like this is not necessary. lossy PGs have been inserted to appl db by refreshPgsForPort

Reply:
Removed.